docs/csharp/async.md

```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
+* `async void` methods are difficult to test.
+* `async void` methods can cause bad side effects if the caller isn't expecting them to be async.
 
 * **Tread carefully when using async lambdas in LINQ expressions**
 
-Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you’re not expecting it to. The introduction of blocking tasks into this can easily result in a deadlock if not written correctly. Additionally, the nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code. Async and LINQ are powerful, but should be used together as carefully and clearly as possible.
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to. The introduction of blocking tasks into this can easily result in a deadlock if not written correctly. Additionally, the nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code. Async and LINQ are powerful, but should be used together as carefully and clearly as possible.
 
 * **Write code that awaits Tasks in a non-blocking manner**
 
-Blocking the current thread as a means to wait for a Task to complete can result in deadlocks and blocked context threads, and can require significantly more complex error-handling. The following table provides guidance on how to deal with waiting for Tasks in a non-blocking way:
+Blocking the current thread as a means to wait for a `Task` to complete can result in deadlocks and blocked context threads, and can require more complex error-handling. The following table provides guidance on how to deal with waiting for Tasks in a non-blocking way:
+
+| Use this...          | Instead of this...           | When wishing to do this                    |
+|----------------------|------------------------------|--------------------------------------------|
+| `await`              | `Task.Wait` or `Task.Result` | Retrieving the result of a background task |
+| `await Task.WhenAny` | `Task.WaitAny`               | Waiting for any task to complete           |
+| `await Task.WhenAll` | `Task.WaitAll`               | Waiting for all tasks to complete          |
+| `await Task.Delay`   | `Thread.Sleep`               | Waiting for a period of time               |
+
+* **Consider using** `ValueTask` **where possible**
+
+Returning a `Task` object from async methods can introduce performance bottlenecks in certain paths. `Task` is a reference type, so using it means allocating an object. In cases where a method declared with the `async` modifier returns a cached result, or completes synchronously, the extra allocations can become a significant time cost in performance critical sections of code. It can become costly if those allocations occur in tight loops. For more information, see [generalized async return types](whats-new/csharp-7.md#generalized-async-return-types).
```

Comment:
```suggestion
Returning a `Task` object from async methods can introduce performance bottlenecks in certain paths. `Task` is a reference type, so using it means allocating an object. In cases where a method declared with the `async` modifier returns a cached result or completes synchronously, the extra allocations can become a significant time cost in performance critical sections of code. It can become costly if those allocations occur in tight loops. For more information, see [generalized async return types](whats-new/csharp-7.md#generalized-async-return-types).
```